src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/TempFileTextBufferManager.cs

```diff
@@ -0,0 +1,182 @@
+            IMsBuildAccessor msbuildAccessor,
+            IVsEditorAdaptersFactoryService editorAdaptersService,
+            ITextDocumentFactoryService textDocumentService,
+            IVsShellUtilitiesHelper shellUtilities,
+            IFileSystem fileSystem,
+            IProjectThreadingService threadingService,
+            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider) :
+#pragma warning disable IDE0030 // Use null propagation https://github.com/dotnet/roslyn/issues/161
+            base(threadingService != null ? threadingService.JoinableTaskContext : throw new ArgumentNullException(nameof(threadingService)))
+#pragma warning restore IDE0030 // Use null propagation
+        {
+            Requires.NotNull(unconfiguredProject, nameof(unconfiguredProject));
+            Requires.NotNull(msbuildAccessor, nameof(msbuildAccessor));
+            Requires.NotNull(editorAdaptersService, nameof(editorAdaptersService));
+            Requires.NotNull(textDocumentService, nameof(textDocumentService));
+            Requires.NotNull(shellUtilities, nameof(shellUtilities));
+            Requires.NotNull(fileSystem, nameof(fileSystem));
+            Requires.NotNull(serviceProvider, nameof(serviceProvider));
+
+            _unconfiguredProject = unconfiguredProject;
+            _msbuildAccessor = msbuildAccessor;
+            _editorAdaptersService = editorAdaptersService;
+            _textDocumentService = textDocumentService;
+            _shellUtilities = shellUtilities;
+            _fileSystem = fileSystem;
+            _threadingService = threadingService;
+            _serviceProvider = serviceProvider;
+        }
+
+        public string FilePath { get; set; }
+
+        public Task InitializeBufferAsync() => InitializeAsync();
+
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            FilePath = GetTempFileName(Path.GetFileName(_unconfiguredProject.FullPath));
+
+            var projectXml = await _msbuildAccessor.GetProjectXmlAsync().ConfigureAwait(false);
+            _fileSystem.WriteAllText(FilePath, projectXml, await _unconfiguredProject.GetFileEncodingAsync().ConfigureAwait(false));
+        }
+
+        protected override Task DisposeCoreAsync(bool initialized)
+        {
+            _fileSystem.RemoveDirectory(Path.GetDirectoryName(FilePath), true);
+            return Task.CompletedTask;
+        }
+
+        public async Task ResetBufferAsync()
+        {
+            var projectXml = await _msbuildAccessor.GetProjectXmlAsync().ConfigureAwait(false);
+
+            await _threadingService.SwitchToUIThread();
+
+            // We compare the text we want to write with the text currently in the buffer, ignoring whitespace. If they're
+            // the same, then we don't write anything.
+            var normalizedExistingText = _whitespaceRegex.Replace(await ReadBufferXmlAsync().ConfigureAwait(true), "");
```

Comment:
Why do we do all this normalization on the UI thread? This cannot be cheap on large projects.